### PR TITLE
Add player GUID and bubble hearth incidents to export string

### PR DESCRIPTION
**Pull request by RobertB4**
codesandbox with the deconversion functions updated:
https://codesandbox.io/s/optimistic-forest-fotoi?file=/src/index.ts

## Potential problems
The export string gets linearly longer the more data we save. At some point we might have to look into compressing the data. For now, the current approach is good enough.

## Screenshot
WoWScrnShot_102721_184344

### CODE REVIEW

File: Hardcore.lua [hunk 1]
```diff
@@ -152,12 +154,15 @@ function Hardcore:PLAYER_LOGIN()
...
 	self:RegisterEvent("UNIT_SPELLCAST_SUCCEEDED")
 
-	if ( Hardcore_Character.time_tracked == nil ) then
-		Hardcore_Character.time_tracked = 0
-	end
-
-	if Hardcore_Character.deaths == nil then
-		Hardcore_Character.deaths = 0
+	-- different guid + lvl1 means new character with the same name
+	if Hardcore_Character.guid ~= PLAYER_GUID and PLAYER_LEVEL == 1 then
```
[the remark this answers is not in the document]
Comment (aaronma37) [reply Hardcore.lua:158]: Alternatively, we could store in hardcore_Character indexed by guid

File: Hardcore.lua [hunk 2]
```diff
@@ -27,6 +27,7 @@ Hardcore_Settings = {
 
 --[[ Character saved variables ]]--
 Hardcore_Character = {
```
Comment (blenderer) [inline Hardcore.lua:29]: Perhaps we should store all of their characters, not just one right?

File: Hardcore.lua [hunk 3]
```diff
@@ -27,6 +27,7 @@ Hardcore_Settings = {
...
+	guid = "",
```
[the remark this answers is not in the document]
Comment (RobertB4) [reply Hardcore.lua:30]: Yes, to compare the guid of the current character to the cached value. This is needed when a player deleted their character and created a new character with the same name. I figured the best solution to compare if 2 characters are the same is by comparing the guid.

Happy to change it up if you have a better solution :)

File: Hardcore.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (RobertB4) [reply Hardcore.lua:158]: Do we still need the data of deleted characters?

[the remark this answers is not in the document]
Comment (RobertB4) [reply Hardcore.lua:158]: > What happens if the player forgets to activate the addon when creating a character and advances past 1?

Good call, the addon will probably be buggy in this case. I will try to think of a solution.

**Review by blenderer — APPROVED** [review]

File: Hardcore.lua [hunk 4]
```diff
@@ -152,12 +154,15 @@ function Hardcore:PLAYER_LOGIN()
...
 	self:RegisterEvent("UNIT_SPELLCAST_SUCCEEDED")
 
-	if ( Hardcore_Character.time_tracked == nil ) then
-		Hardcore_Character.time_tracked = 0
-	end
-
-	if Hardcore_Character.deaths == nil then
-		Hardcore_Character.deaths = 0
+	-- different guid + lvl1 means new character with the same name
+	if Hardcore_Character.guid ~= PLAYER_GUID then
```
Comment (RobertB4) [inline Hardcore.lua:158]: Removed the check for PLAYER_LEVEL == 1 from this condition.
This should now work for every case.

First character with name:
Hardcore_Character.guid is nil, PLAYER_GUID is the guid of the character.

New character with same name (old character was deleted):
Hardcore_Character.guid has a set value that is different from PLAYER_GUID

Character advances past level 1 before turning on the addon:
There no longer is a check for level, so this is not an issue.

**Review by blenderer — APPROVED** [review]
Good stuff!

**Review by zdeyn — APPROVED** [review]
I don't see any major problems here - worse-case, post-integration tweaking

**Pull request merged**